File: web/lib/geography.ts
/**
 * Geografia dell'Isola dei Tre Venti — mirror TypeScript del modulo Python
 * in `tools/geography.py`. Stessa semantica, stessi preset, stesso algoritmo.
 *
 * Sorgente di verità: `world/geography/*.json` (fuori da `web/`).
 *
 * Convenzione coordinate: Three.js. Origine (0,0) = Albero Vecchio.
 * est=+x, sud=+z, su=+y. Unità nel JSON: metri.
 *
 * Questo file è client-safe (nessun import Node). La funzione di caricamento
 * da filesystem sta in `geography-load.ts` (server-only).
 */

// ---------------------------------------------------------------------------
// Preset velocità canoniche — m/s
// ---------------------------------------------------------------------------

export const WALKING_SPEEDS_MPS = {
  adult_walk: 1.2,
  child_walk: 0.8,
  bartolo_walk: 0.4,
  child_run: 2.0,
} as const;

export type SpeedPreset = keyof typeof WALKING_SPEEDS_MPS;

// ---------------------------------------------------------------------------
// Tipi dati (shape dei JSON in world/geography/)
// ---------------------------------------------------------------------------

export type Coords = {
  x: number;
  z: number;
  y: number;
};

export type Quartiere = {
  id: string;
  label: string;
  center: { x: number; z: number };
  radius_m: number;
  elevation_m: number;
  color: string;
  descrizione: string;
};

/**
 * Shape di `world/geography/island.json` (bounds + 5 quartieri).
 * Storicamente chiamato `IslandGeography`; rinominato in `IslandGeographyData`
 * per evitare collisione con la classe di query omonima.
 */
export type IslandGeographyData = {
  version: string;
  island: {
    bounds: {
      width_m: number;
      depth_m: number;
      unit_scale_three: number;
    };
    shape: "oval" | "circle" | "rect";
  };
  quartieri: Quartiere[];
  notes?: string;
};

export type Location = {
  id: string;
  nome_canonico: string;
  categoria: string;
  quartiere: string | null;
  coords: Coords;
  descrizione_breve: string;
  porta_socchiusa: boolean;
};

export type LocationRaw = Location & {
  fonti: string[];
};

export type LocationsDoc = {
  version: string;
  notes?: string;
  locations: LocationRaw[];
};

export type Edge = {
  from: string;
  to: string;
  via: string | null;
  distance_m: number;
  difficulty: "easy" | "steep" | "sandy";
};

export type NamedRoute = {
  id: string;
  nome_canonico?: string;
  from_quartiere: string;
  to_quartiere: string;
};

export type PathsDoc = {
  version: string;
  notes?: string;
  named_routes: NamedRoute[];
  edges: Edge[];
};

// ---------------------------------------------------------------------------
// Helpers scena
// ---------------------------------------------------------------------------

/**
 * Convert a world coordinate (meters) to Three.js scene units.
 * 1 Three.js unit = 100 m (unit_scale_three). Three.js Y is up;
 * our canonical x = east, z = south (positive), y = elevation.
 */
export function metersToUnits(meters: number, scale = 100): number {
  return meters / scale;
}

/**
 * Palette — mirror of legacy/index.html CSS variables so the 3D scene
 * stays coherent with the paper-map prototype.
 */
export const PALETTE = {
  cream: "#f5ead5",
  creamShade: "#ebdebe",
  creamDeep: "#e0ceaa",
  terracotta: "#c8532c",
  terracottaDeep: "#9c3d1e",
  olive: "#5c6b3b",
  oliveLight: "#8a9a6b",
  sea: "#1e5d6e",
  seaLight: "#4a8595",
  seaDeep: "#0f3a48",
  ink: "#2a2319",
  warmBrown: "#6b4423",
  sun: "#e8a547",
  sand: "#d8bd8a",
} as const;

// ---------------------------------------------------------------------------
// API interrogabile (mirror di tools/geography.py::IslandGeography)
// ---------------------------------------------------------------------------

function euclidean(a: Coords, b: Coords): number {
  const dx = a.x - b.x;
  const dz = a.z - b.z;
  const dy = a.y - b.y;
  return Math.sqrt(dx * dx + dz * dz + dy * dy);
}

/**
 * Mini-heap binario (stdlib TS non include PriorityQueue).
 * Sufficiente per Dijkstra su poche decine di nodi.
 */
class MinHeap<T> {
  private items: Array<{ key: number; val: T }> = [];

  push(key: number, val: T): void {
    this.items.push({ key, val });
    this.bubbleUp(this.items.length - 1);
  }

  pop(): { key: number; val: T } | undefined {
    if (this.items.length === 0) return undefined;
    const top = this.items[0];
    const last = this.items.pop()!;
    if (this.items.length > 0) {
      this.items[0] = last;
      this.sinkDown(0);
    }
    return top;
  }

  get size(): number {
    return this.items.length;
  }

  private bubbleUp(i: number): void {
    while (i > 0) {
      const p = Math.floor((i - 1) / 2);
      if (this.items[p].key <= this.items[i].key) break;
      [this.items[p], this.items[i]] = [this.items[i], this.items[p]];
      i = p;
    }
  }

  private sinkDown(i: number): void {
    const n = this.items.length;
    while (true) {
      const l = 2 * i + 1;
      const r = 2 * i + 2;
      let smallest = i;
      if (l < n && this.items[l].key < this.items[smallest].key) smallest = l;
      if (r < n && this.items[r].key < this.items[smallest].key) smallest = r;
      if (smallest === i) break;
      [this.items[smallest], this.items[i]] = [
        this.items[i],
        this.items[smallest],
      ];
      i = smallest;
    }
  }
}

export class IslandGeography {
  private readonly island: IslandGeographyData;
  private readonly locs: Map<string, Location>;
  private readonly edgesList: Edge[];
  private readonly adj: Map<string, Array<[string, number]>>;
  private readonly routes: NamedRoute[];

  constructor(
    island: IslandGeographyData,
    locations: LocationsDoc,
    paths: PathsDoc,
  ) {
    this.island = island;
    this.locs = new Map();
    for (const raw of locations.locations) {
      const loc: Location = {
        id: raw.id,
        nome_canonico: raw.nome_canonico,
        categoria: raw.categoria,
        quartiere: raw.quartiere,
        coords: {
          x: Number(raw.coords.x),
          z: Number(raw.coords.z),
          y: Number(raw.coords.y),
        },
        descrizione_breve: raw.descrizione_breve,
        porta_socchiusa: Boolean(raw.porta_socchiusa),
      };
      this.locs.set(loc.id, loc);
    }
    this.edgesList = paths.edges.map((e) => ({ ...e }));
    this.adj = new Map();
    for (const id of this.locs.keys()) {
      this.adj.set(id, []);
    }
    for (const e of this.edgesList) {
      if (!this.adj.has(e.from)) this.adj.set(e.from, []);
      if (!this.adj.has(e.to)) this.adj.set(e.to, []);
      this.adj.get(e.from)!.push([e.to, e.distance_m]);
      this.adj.get(e.to)!.push([e.from, e.distance_m]);
    }
    this.routes = paths.named_routes.map((r) => ({ ...r }));
  }

  /** Costruisce da JSON già letti (client-side, test, ecc.) */
  static fromData(
    island: IslandGeographyData,
    locations: LocationsDoc,
    paths: PathsDoc,
  ): IslandGeography {
    return new IslandGeography(island, locations, paths);
  }

  // ---------------- lookup ----------------

  location(id: string): Location {
    const l = this.locs.get(id);
    if (!l) throw new Error(`Location sconosciuta: ${id}`);
    return l;
  }

  allLocations(): Location[] {
    return Array.from(this.locs.values()).sort((a, b) =>
      a.id.localeCompare(b.id),
    );
  }

  quartieri(): Quartiere[] {
    return this.island.quartieri.map((q) => ({ ...q }));
  }

  islandBounds(): IslandGeographyData["island"]["bounds"] {
    return { ...this.island.island.bounds };
  }

  edges(): Edge[] {
    return this.edgesList.map((e) => ({ ...e }));
  }

  namedRoutes(): NamedRoute[] {
    return this.routes.map((r) => ({ ...r }));
  }

  // ---------------- distanza e tempo ----------------

  distance(a: string, b: string): number {
    return euclidean(this.location(a).coords, this.location(b).coords);
  }

  pathDistance(a: string, b: string): number {
    const route = this.path(a, b);
    if (route.length < 2) return this.distance(a, b);
    let total = 0;
    for (let i = 0; i < route.length - 1; i++) {
      const u = route[i];
      const v = route[i + 1];
      const nbrs = this.adj.get(u) ?? [];
      let step: number | null = null;
      for (const [nb, d] of nbrs) {
        if (nb === v) {
          step = d;
          break;
        }
      }
      total += step ?? this.distance(u, v);
    }
    return total;
  }

  walkingTime(a: string, b: string, speed: SpeedPreset = "adult_walk"): number {
    if (!(speed in WALKING_SPEEDS_MPS)) {
      throw new Error(`Preset velocità sconosciuto: ${speed}`);
    }
    const mps = WALKING_SPEEDS_MPS[speed];
    const meters = this.pathDistance(a, b);
    return meters / mps / 60; // minuti
  }

  // ---------------- pathfinding ----------------

  /**
   * Dijkstra sulla rete sentieri. Se a==b → [a]. Se non connessi →
   * cammino virtuale [a,b] (parità con l'implementazione Python).
   */
  path(a: string, b: string): string[] {
    if (!this.locs.has(a)) throw new Error(`Location sconosciuta: ${a}`);
    if (!this.locs.has(b)) throw new Error(`Location sconosciuta: ${b}`);
    if (a === b) return [a];

    const dist = new Map<string, number>();
    const prev = new Map<string, string | null>();
    dist.set(a, 0);
    prev.set(a, null);
    const pq = new MinHeap<string>();
    pq.push(0, a);
    const visited = new Set<string>();

    while (pq.size > 0) {
      const entry = pq.pop()!;
      const u = entry.val;
      const d = entry.key;
      if (visited.has(u)) continue;
      visited.add(u);
      if (u === b) break;
      const nbrs = [...(this.adj.get(u) ?? [])].sort((x, y) => {
        if (x[0] < y[0]) return -1;
        if (x[0] > y[0]) return 1;
        return 0;
      });
      for (const [v, w] of nbrs) {
        const nd = d + w;
        const current = dist.has(v) ? dist.get(v)! : Number.POSITIVE_INFINITY;
        if (nd < current) {
          dist.set(v, nd);
          prev.set(v, u);
          pq.push(nd, v);
        }
      }
    }

    if (!dist.has(b)) {
      // non connesso: cammino virtuale
      return [a, b];
    }

    const route: string[] = [];
    let cur: string | null = b;
    while (cur !== null && cur !== undefined) {
      route.push(cur);
      cur = prev.get(cur) ?? null;
    }
    route.reverse();
    return route;
  }

  // ---------------- query spaziali ----------------

  nearby(loc: string, radiusM = 500): string[] {
    const center = this.location(loc).coords;
    const out: Array<{ d: number; id: string }> = [];
    for (const other of this.locs.values()) {
      if (other.id === loc) continue;
      const d = euclidean(center, other.coords);
      if (d <= radiusM) out.push({ d, id: other.id });
    }
    out.sort((x, y) => {
      if (x.d !== y.d) return x.d - y.d;
      return x.id.localeCompare(y.id);
    });
    return out.map((r) => r.id);
  }

  quartiereOf(loc: string): string | null {
    return this.location(loc).quartiere;
  }

  elevation(loc: string): number {
    return this.location(loc).coords.y;
  }

  /**
   * Euristica approssimata di visibilità (non true line-of-sight).
   * Vedi `tools/geography.py::visible_from` per il contratto semantico.
   */
  visibleFrom(loc: string, elevationBonusM = 0): string[] {
    const center = this.location(loc);
    const ea = center.coords.y + elevationBonusM;
    const out: Array<{ d: number; id: string }> = [];
    for (const other of this.locs.values()) {
      if (other.id === loc) continue;
      if (other.porta_socchiusa) continue;
      const d = euclidean(center.coords, other.coords);
      let visible = false;
      if (ea > other.coords.y) {
        visible = true;
      } else if (center.coords.y >= 100 && d <= 3000) {
        visible = true;
      }
      if (visible) out.push({ d, id: other.id });
    }
    out.sort((x, y) => {
      if (x.d !== y.d) return x.d - y.d;
      return x.id.localeCompare(y.id);
    });
    return out.map((r) => r.id);
  }
}

File: tools/geography.py
"""Geography module for L'Isola dei Tre Venti.

Source of truth: world/geography/*.json
- island.json    — bounds + 5 quartieri (letto da web app esistente)
- locations.json — luoghi canonici con coordinate
- paths.json     — rete sentieri per pathfinding

Convenzione coordinate: Three.js. Origine (0,0) = Albero Vecchio.
est=+x, sud=+z, su=+y. Unità nel JSON: metri.

Velocità di percorrenza canoniche (preset nominati, in m/s):
- adult_walk    1.2   abitanti adulti sui sentieri
- child_walk    0.8   i tre fratelli
- bartolo_walk  0.4   tartaruga fuori dall'acqua
- child_run     2.0   gioco/fuga

Vincoli canonici:
- Nessun framework EAR, nessun Pattern A, nessun seed/callback.
- Solo geografia osservabile.
- Quartieri identificati da id tecnici (centro/forno/pontile/orti/montagne);
  MAI label 'Fuoco/Acqua/Terra/Aria' in output user-facing.
"""
from __future__ import annotations

import heapq
import json
import math
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal

WALKING_SPEEDS_MPS: dict[str, float] = {
    "adult_walk": 1.2,
    "child_walk": 0.8,
    "bartolo_walk": 0.4,
    "child_run": 2.0,
}

SpeedPreset = Literal["adult_walk", "child_walk", "bartolo_walk", "child_run"]


@dataclass(frozen=True)
class Coords:
    x: float
    z: float
    y: float


@dataclass(frozen=True)
class Location:
    id: str
    nome_canonico: str
    categoria: str
    quartiere: str | None
    coords: Coords
    descrizione_breve: str
    porta_socchiusa: bool


@dataclass(frozen=True)
class Edge:
    from_id: str
    to_id: str
    via: str | None
    distance_m: float
    difficulty: str


def _euclidean(a: Coords, b: Coords) -> float:
    return math.sqrt(
        (a.x - b.x) ** 2 + (a.z - b.z) ** 2 + (a.y - b.y) ** 2
    )


def _resolve_data_dir(data_dir: str | Path | None) -> Path:
    """Trova la cartella world/geography. Accetta un path assoluto,
    un path relativo alla repo root, oppure None (autodetect)."""
    if data_dir is None:
        # autodetect: sali da questo file fino alla repo root
        here = Path(__file__).resolve()
        repo_root = here.parent.parent
        return repo_root / "world" / "geography"
    p = Path(data_dir)
    if p.is_absolute():
        return p
    # relativo a cwd o repo root: prova entrambi, preferisci cwd
    cwd_p = Path.cwd() / p
    if cwd_p.exists():
        return cwd_p
    here = Path(__file__).resolve()
    repo_root = here.parent.parent
    return repo_root / p


class IslandGeography:
    """API interrogabile sulla geografia dell'isola.

    Stesso dato (world/geography/*.json) consumato da:
    - autori AI in chat narrativa (tramite tools/geo_query.py CLI)
    - autori umani da terminale (CLI)
    - web app (mirror TypeScript in web/lib/geography.ts)
    """

    def __init__(self, data_dir: str | Path | None = None) -> None:
        base = _resolve_data_dir(data_dir)
        with (base / "island.json").open("r", encoding="utf-8") as fh:
            self._island: dict[str, Any] = json.load(fh)
        with (base / "locations.json").open("r", encoding="utf-8") as fh:
            loc_doc: dict[str, Any] = json.load(fh)
        with (base / "paths.json").open("r", encoding="utf-8") as fh:
            path_doc: dict[str, Any] = json.load(fh)

        self._locations: dict[str, Location] = {}
        for raw in loc_doc["locations"]:
            c = raw["coords"]
            loc = Location(
                id=raw["id"],
                nome_canonico=raw["nome_canonico"],
                categoria=raw["categoria"],
                quartiere=raw.get("quartiere"),
                coords=Coords(float(c["x"]), float(c["z"]), float(c["y"])),
                descrizione_breve=raw.get("descrizione_breve", ""),
                porta_socchiusa=bool(raw.get("porta_socchiusa", False)),
            )
            self._locations[loc.id] = loc

        self._edges: list[Edge] = []
        self._adj: dict[str, list[tuple[str, float]]] = {
            lid: [] for lid in self._locations
        }
        for raw in path_doc["edges"]:
            e = Edge(
                from_id=raw["from"],
                to_id=raw["to"],
                via=raw.get("via"),
                distance_m=float(raw["distance_m"]),
                difficulty=raw.get("difficulty", "easy"),
            )
            self._edges.append(e)
            # grafo non direzionato
            self._adj.setdefault(e.from_id, []).append((e.to_id, e.distance_m))
            self._adj.setdefault(e.to_id, []).append((e.from_id, e.distance_m))

        self._named_routes: list[dict[str, Any]] = list(
            path_doc.get("named_routes", [])
        )

    # ------------------------------------------------------------------
    # Lookup base
    # ------------------------------------------------------------------

    def location(self, id: str) -> Location:
        if id not in self._locations:
            raise KeyError(f"Location sconosciuta: {id!r}")
        return self._locations[id]

    def all_locations(self) -> list[Location]:
        # ordinamento deterministico per id
        return [self._locations[k] for k in sorted(self._locations)]

    def quartieri(self) -> list[dict[str, Any]]:
        return list(self._island["quartieri"])

    def island_bounds(self) -> dict[str, Any]:
        return dict(self._island["island"]["bounds"])

    # ------------------------------------------------------------------
    # Distanze e tempi
    # ------------------------------------------------------------------

    def distance(self, a: str, b: str) -> float:
        """Distanza in linea retta (3D euclidea) in metri."""
        la = self.location(a)
        lb = self.location(b)
        return _euclidean(la.coords, lb.coords)

    def path_distance(self, a: str, b: str) -> float:
        """Distanza lungo la rete sentieri (somma archi Dijkstra).
        Se i due nodi non sono connessi, fallback alla linea retta."""
        route = self.path(a, b)
        if len(route) < 2:
            return self.distance(a, b)
        total = 0.0
        for i in range(len(route) - 1):
            u, v = route[i], route[i + 1]
            # cerca l'arco (u,v) nella adiacenza
            step = None
            for nb, d in self._adj.get(u, []):
                if nb == v:
                    step = d
                    break
            if step is None:
                # non dovrebbe accadere: path() ritorna un cammino consistente
                step = self.distance(u, v)
            total += step
        return total

    def walking_time(
        self, a: str, b: str, speed: SpeedPreset = "adult_walk"
    ) -> float:
        """Tempo di percorrenza in minuti, lungo la rete sentieri
        (fallback: linea retta se i nodi non sono connessi).

        Il preset controlla solo la velocità; non modella la difficoltà
        degli archi (steep/sandy) — quella è informativa sull'arco.
        """
        if speed not in WALKING_SPEEDS_MPS:
            raise ValueError(
                f"Preset velocità sconosciuto: {speed!r}. "
                f"Valori ammessi: {sorted(WALKING_SPEEDS_MPS)}"
            )
        mps = WALKING_SPEEDS_MPS[speed]
        meters = self.path_distance(a, b)
        return (meters / mps) / 60.0  # secondi -> minuti

    # ------------------------------------------------------------------
    # Pathfinding
    # ------------------------------------------------------------------

    def path(self, a: str, b: str) -> list[str]:
        """Sequenza di location id per il percorso più breve nella rete
        sentieri (Dijkstra sui pesi distance_m di paths.json).

        Se a o b sono sconosciuti: KeyError.
        Se non esiste cammino nella rete: ritorna [a, b] (edge virtuale
        in linea retta — usato dai punti extra-rete come isole offshore).
        """
        if a not in self._locations:
            raise KeyError(f"Location sconosciuta: {a!r}")
        if b not in self._locations:
            raise KeyError(f"Location sconosciuta: {b!r}")
        if a == b:
            return [a]

        # Dijkstra
        dist: dict[str, float] = {a: 0.0}
        prev: dict[str, str | None] = {a: None}
        pq: list[tuple[float, str]] = [(0.0, a)]
        visited: set[str] = set()

        while pq:
            d, u = heapq.heappop(pq)
            if u in visited:
                continue
            visited.add(u)
            if u == b:
                break
            for v, w in sorted(self._adj.get(u, [])):
                nd = d + w
                if nd < dist.get(v, math.inf):
                    dist[v] = nd
                    prev[v] = u
                    heapq.heappush(pq, (nd, v))

        if b not in dist:
            # nodi non connessi: ritorna cammino virtuale
            return [a, b]

        # ricostruisci
        route: list[str] = []
        cur: str | None = b
        while cur is not None:
            route.append(cur)
            cur = prev.get(cur)
        route.reverse()
        return route

    # ------------------------------------------------------------------
    # Query spaziali
    # ------------------------------------------------------------------

    def nearby(self, loc: str, radius_m: float = 500) -> list[str]:
        """Id delle location entro radius_m (linea retta) da loc.
        Esclude loc stessa. Ordinate per distanza crescente, poi id."""
        center = self.location(loc)
        out: list[tuple[float, str]] = []
        for other in self._locations.values():
            if other.id == loc:
                continue
            d = _euclidean(center.coords, other.coords)
            if d <= radius_m:
                out.append((d, other.id))
        out.sort(key=lambda r: (r[0], r[1]))
        return [oid for _, oid in out]

    def quartiere_of(self, loc: str) -> str | None:
        """Il quartiere che contiene loc, o None per feature extra-quartiere
        (fiume, guado, isole offshore)."""
        return self.location(loc).quartiere

    def elevation(self, loc: str) -> float:
        return self.location(loc).coords.y

    def visible_from(
        self, loc: str, elevation_bonus_m: float = 0
    ) -> list[str]:
        """Euristica semplice (non line-of-sight reale) di visibilità:

        Una location B è considerata visibile da A se vale almeno una di:
        - elevation(A) + elevation_bonus > elevation(B)
          (A guarda verso il basso o alla stessa quota)
        - elevation(A) >= 100m AND distance(A,B) <= 3000m
          (punto panoramico: orizzonte libero a breve-media distanza)

        Esclude A stessa e le porte_socchiuse (non sono 'nell'isola').
        Ordine: distanza crescente, poi id.

        Approssimazione: non tiene conto di ostacoli intermedi
        (Montagne Gemelle tra osservatore e valle, per esempio).
        """
        center = self.location(loc)
        ea = center.coords.y + elevation_bonus_m
        out: list[tuple[float, str]] = []
        for other in self._locations.values():
            if other.id == loc:
                continue
            if other.porta_socchiusa:
                continue
            d = _euclidean(center.coords, other.coords)
            visible = False
            if ea > other.coords.y:
                visible = True
            elif center.coords.y >= 100 and d <= 3000:
                visible = True
            if visible:
                out.append((d, other.id))
        out.sort(key=lambda r: (r[0], r[1]))
        return [oid for _, oid in out]

    # ------------------------------------------------------------------
    # Introspezione
    # ------------------------------------------------------------------

    def edges(self) -> list[Edge]:
        return list(self._edges)

    def named_routes(self) -> list[dict[str, Any]]:
        return list(self._named_routes)
